### PR TITLE
feat: Add `ABCIMethodLatency` histogram [BLO-791]

### PR DESCRIPTION
**Pull request by g3ndo9kar9**
## In This PR
- I introduce the `ABCIMethodLatency` histogram in `service/metrics/metrics.go` , see [the RFC](https://www.notion.so/skip-protocol/RFC-24-Slinky-Deployment-Metrics-092763260ef5495d8fd93f596bb1547d)
- I instrument Prepare/ProcessProposal
- I add a `README` section for metrics, and a README for `service/metrics`
- I fix the e2e tests to expose the app-metrics port
- misc changes are documented (all bug-fixes)

## Example output
Screen Shot 2024-01-22 at 3 19 42 PM

### CODE REVIEW

File: service/metrics/metrics.go [hunk 1]
```diff
@@ -101,27 +125,20 @@ func NewMetrics() Metrics {
...
-		processProposalTime: prometheus.NewHistogram(prometheus.HistogramOpts{
+		abciMethodLatency: prometheus.NewHistogramVec(prometheus.HistogramOpts{
 			Namespace: AppNamespace,
-			Name:      "oracle_process_proposal_time",
-			Help:      "The time it took for the oracle-specific parts of process proposal",
-			Buckets:   prometheus.ExponentialBuckets(1, 2, 10),
-		}),
+			Name:      "abci_method_latency",
+			Help:      "The time it took for an ABCI method to execute slinky specific logic (in seconds)",
+			Buckets:   []float64{.0001, .0004, .002, .009, .02, .1, .65, 2, 6, 25},
```
Comment (g3ndo9kar9) [inline service/metrics/metrics.go:132]: buckets chosen are in seconds, these are the same buckets used by cometBFTs, abci_method_latency histogram

File: tests/integration/slinky_suite.go [hunk 2]
```diff
@@ -77,7 +77,7 @@ func DefaultOracleConfig(node *cosmos.ChainNode) (
 
 	// Create the metrics config
 	metricsConfig := oracleconfig.MetricsConfig{
-		PrometheusServerAddress: fmt.Sprintf("%s:%s", oracle.HostName(), "8081"),
+		PrometheusServerAddress: fmt.Sprintf("%s:%s", "0.0.0.0", "8081"),
```
Comment (g3ndo9kar9) [inline tests/integration/slinky_suite.go:80]: expose app-metrics to docker-host, as opposed to oracle-host in docker-network

File: .dockerignore [hunk 3]
```diff
@@ -1,2 +1 @@
 go.work*
```
Comment (g3ndo9kar9) [inline .dockerignore:1]: un ignore go.sum so dockerfiles can build

File: abci/proposals/proposals.go [hunk 4]
```diff
@@ -276,6 +270,19 @@ func (h *ProposalHandler) ProcessProposalHandler() sdk.ProcessProposalHandler {
...
+		resp, err := h.processProposalHandler(ctx, req)
+		if err == nil {
+			// record time spent in slinky-specific ProcesssProposal on successful process-proposals
+			h.logger.Info(
+				"recording process proposal time",
+				"duration (seconds)", processDuration.Seconds(),
+			)
+			h.metrics.ObserveABCIMethodLatency(servicemetrics.ProcessProposal, processDuration)
+		}
+		return resp, err
```
Comment (Eric-Warehime) [inline abci/proposals/proposals.go]: The reason I had metrics observation in a defer was to guarantee that we reported a metric no matter what our response was.

The way you have it above will only report in the happy path and not any other result. I think we want to report metrics no matter whether the response is Accept or Reject.

Comment (g3ndo9kar9) [reply abci/proposals/proposals.go]: Ya, I intentionally wanted to only collect latencies for successful requests.

Comment (g3ndo9kar9) [reply abci/proposals/proposals.go]: You'd prefer to report in all cases?

File: service/metrics/metrics.go [hunk 5]
```diff
@@ -57,11 +85,8 @@ type Metrics interface {
 	// AddTickerInclusionStatus increments the counter representing the number of times a ticker was included (or not included) in the last commit.
 	AddTickerInclusionStatus(ticker string, included bool)
 
-	// ObserveProcessProposalTime records the time it took for the oracle-specific parts of process proposal
-	ObserveProcessProposalTime(duration time.Duration)
-
-	// ObservePrepareProposalTime records the time it took for the oracle-specific parts of prepare proposal
-	ObservePrepareProposalTime(duration time.Duration)
+	// ObserveABCIMethodLatency reports the given latency (as a duration), for the given ABCIMethod, and updates the ABCIMethodLatency histogram w/ that value.
+	ObserveABCIMethodLatency(method ABCIMethod, duration time.Duration)
```
Comment (Eric-Warehime) [inline service/metrics/metrics.go:89]: Nice 👍

**Review by Eric-Warehime — APPROVED** [review]

**Pull request merged**